sentinelsat/advanced.py

```diff
@@ -0,0 +1,359 @@
+
+            node_info["size"] = int(info["ContentLength"])
+
+            response = self.session.get(node_info["url"], auth=self.session.auth)
+            _check_scihub_response(response, test_json=False)
+            data = response.content
+            if len(data) != node_info["size"]:
+                raise SentinelAPIError("File corrupt: data length do not match")
+
+            if path:
+                os.makedirs(os.path.dirname(path), exist_ok=True)
+                with open(path, "wb") as fd:
+                    fd.write(data)
+
+        return node_info, data
+
+    def _dataobj_to_node_info(self, dataobj_info, product_info):
+        path = dataobj_info["href"]
+        if path.startswith("./"):
+            path = path[2:]
+
+        node_info = product_info.copy()
+        node_info["url"] = self._path_to_url(product_info, path, "value")
+        node_info["size"] = dataobj_info["size"]
+        node_info["md5"] = dataobj_info["md5"]
+        node_info["node_path"] = dataobj_info["href"]
+        # node_info["parent"] = product_info
+
+        return node_info
+
+    def _filter_nodes(self, manifest, product_info, nodefilter=None):
+        if nodefilter is None:
+            nodefilter = self.nodefilter
+
+        nodes = {}
+        xmldoc = etree.parse(manifest)
+        data_obj_section_elem = xmldoc.find("dataObjectSection")
+        for elem in data_obj_section_elem.iterfind("dataObject"):
+            dataobj_info = _xml_to_dataobj_info(elem)
+            node_info = self._dataobj_to_node_info(dataobj_info, product_info)
+            if nodefilter is not None and not nodefilter(node_info):
+                continue
+            node_path = node_info["node_path"]
+            nodes[node_path] = node_info
+        return nodes
+
+    def download(self, id, directory_path=".", checksum=True):
+        """Download a product.
+
+        Uses the filename on the server for the downloaded files, e.g.
+        "S1A_EW_GRDH_1SDH_20141003T003840_20141003T003920_002658_002F54_4DD1.SAFE/manifest.safe".
+
+        Incomplete downloads are continued and complete files are skipped.
+
+        Parameters
+        ----------
+        id : string
+            UUID of the product, e.g. 'a8dd0cfd-613e-45ce-868c-d79177b916ed'
+        directory_path : string, optional
+            Where the file will be downloaded
+        checksum : bool, optional
+            If True, verify the downloaded file's integrity by checking its MD5 checksum.
+            Throws InvalidChecksumError if the checksum does not match.
+            Defaults to True.
+
+        Returns
+        -------
+        product_info : dict
+            Dictionary containing the product's info from get_product_info() as well as
+            the path on disk.
+
+        Raises
+        ------
+        InvalidChecksumError
+            If the MD5 checksum does not match the checksum on the server.
+        """
+        if self.nodefilter is None:
```

Comment:
Yes I'm don't like too much that solution too. And passing an additional parameter is an option that I have also considered.

The main reason why I have chosen to use the solution with the ``.nodefilter`` attribute is that it allows to not change the original ``sentinel.py`` module. Right now I can use the new features just having a local copy of the ``advanced.py`` module and a standard (conda) installation of sentinelsat.

Now that I know that you are interested in having the new features in sentinelsat I'm totally in favour of changing the approach and passing additional parameters.

Anyway there is a practical question. If we decide to pass around the ``nodefilter`` parameter that we probably need to move all the new stuff into the main ``SentinelAPI`` class, otherwise we will need to override basically all the download related methods.

Another option would be to modify ``SentinelAPI`` download methods to accept an extra ``**kwargs`` and keep the new stuff in a separate class.